sdks/python/apache_beam/io/gcp/spannerio.py

```diff
@@ -0,0 +1,558 @@
+        database will construct an instance of BurstyPool.
+      read_timestamp: (optional) An instance of the `datetime.datetime` object
+        to execute all reads at the given timestamp.
+      exact_staleness: (optional) An instance of `datetime.timedelta` object
+        to execute all reads at a timestamp that is exact_staleness old.
+      credentials: (optional) The authorization credentials to attach to
+        requests. These credentials identify this application to the service.
+        If none are specified, the client will attempt to ascertain
+        the credentials from the environment.
+      sql: (optional) SQL query statement
+      params: (optional) Values for parameter replacement. Keys must match the
+        names used in sql.
+      param_types: (optional) maps explicit types for one or more param values;
+        required if params are passed.
+      table: (optional) Name of the table from which to fetch data.
+      columns: (optional) List of names of columns to be retrieved; required if
+        the table is passed.
+      index: (optional) name of index to use, rather than the table's primary
+        key.
+      keyset: (optional) keys / ranges identifying rows to be retrieved.
+      read_operations: (optional) List of the objects of :class:`_ReadOperation`
+        to perform read all.
+      transaction: (optional) PTransform of the :meth:`_create_transaction` to
+        perform naive read on cloud spanner.
+    """
+    self._configuration = _BeamSpannerConfiguration(
+        project=project_id, instance=instance_id, database=database_id,
+        credentials=credentials, pool=pool,
+        snapshot_read_timestamp=read_timestamp,
+        snapshot_exact_staleness=exact_staleness
+    )
+
+    self._read_operations = read_operations
+    self._transaction = transaction
+
+    if self._read_operations is None:
+      if table is not None:
+        if columns is None:
+          raise ValueError("Columns are required with the table name.")
+        self._read_operations = [_ReadOperation.table(
+            table=table, columns=columns, index=index, keyset=keyset)]
+      elif sql is not None:
+        self._read_operations = [_ReadOperation.query(
+            sql=sql, params=params, param_types=param_types)]
+
+  def expand(self, pbegin):
+    if self._read_operations is not None and isinstance(pbegin,
+                                                        PBegin):
+      pcoll = pbegin.pipeline | Create(self._read_operations)
+    elif not isinstance(pbegin, PBegin):
+      if self._read_operations is not None:
+        raise ValueError("Read operation in the constructor only works with "
+                         "the root of the pipeline.")
+      pcoll = pbegin
+    else:
+      raise ValueError("Spanner required read operation, sql or table "
+                       "with columns.")
+
+    if self._transaction is None:
+      # reading as batch read
+      p = (pcoll
```

Comment:
I'm not sure why we are forking here based on whether a transaction is provided or not. Seems like, in Java version, we use transactions for both native and batch versions of the read transform: https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java#L420

Reply:
If user does not provide the transaction, we use BatchSnapshot to generate query batches. The reason we can not use the transaction here because `google.cloud.spanner_v1.database.BatchSnapshot.generate_read_batches` requires `snapshot` with `multi_use=True` and uses a private method to create snapshot instance (`google.cloud.spanner_v1.database.BatchSnapshot._get_snapshot`).

On the other hand, `create_trasaction` simply returns the transaction_id and session_id which we reconstruct in the naive read using `google.cloud.spanner_v1.database.BatchSnapshot.from_dict` with no way to set `multi_use=True` Ref: https://github.com/googleapis/google-cloud-python/blob/master/spanner/google/cloud/spanner_v1/database.py#L651

Reply:
So not using a transaction may offer better performance ? We should clarify this in the documentation.